training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+#### Try It Yourself 2: Array Absolute Value
+Problem:
+Write a `proc` using a ternary to take an `int array`, `A`, 
+and return an array where index `i` is the absolute value of `A[i]`
+
+Input: 
+`[-3, 7, 0, -4, 12]`
+
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+proc arrayAbsVal(A) {
+  return [a in A] if a >= 0 then a else -a;
+}
+```
+</details>
+Expected Output:
+
+```console
+3 7 0 4 12
+```
+
+<a id="introspection"></a>
+## Introspection
+Introspection is determining properties of a function argument at runtime.
+This is often used to determine the type and/or domain of a function argument.
+Introspection can avoid duplicating a `proc` for multiple types when none of the logic has changed.
+Using introspection will result in a [generic](https://chapel-lang.org/docs/language/spec/generics.html#generics) function.
+
+The syntax for this is:
+```Chapel
+proc foo(array: [?D] ?t, value: t, array2: [?D2] ?t2) {
+  // using `D` and `t` in this proc will refer to `array`'s domain and type respectively
+  // `value` has the same type as `array` so there's no need for the `?`
+  // `D2` and `t2` refer to the domain and type of `array2`
+}
+```
+<a id="ex12"></a>
+#### Example 12: Introspection
+If we adapt our absolute value example to use introspection,
+we can use the same `proc` for `real array`s and `int array`s.
+```Chapel
+proc absoluteVal(a: [?D] ?t): [D] t {
+  return [i in D] if a[i] >= 0 then a[i] else -a[i];
+}
+
+var r: [0..#5] real = [-3.14, 7:real, 0.0, INFINITY, -INFINITY];
+writeln(absoluteVal(r));
+
+var i: [0..#5] int = [-3, 7, 0, -4, 12];
+writeln(absoluteVal(i));
+```
+```console
+3.14 7.0 0.0 inf inf
+3 7 0 4 12
+```
+
+<a id="promotion"></a>
+## Promotion
+[Promotion](https://chapel-lang.org/docs/users-guide/datapar/promotion.html) is a way to obtain data parallelism implicitly.
+
+A function or operation that operates on a type can automatically work on an array of that type
+by essentially applying the function to every element of the array.
+
+<a id="ex13"></a>
+#### Example 13: Promotion
+Returning to the factorial example, our `proc` can operate on an `int array` even though it's only defined to accept an `int`.
+```chapel
+proc factorial(n: int) {
+  return * reduce [i in 1..#n] i;
+}
+writeln(factorial(5));
+writeln(factorial([1, 2, 3, 4, 5]));
+```
+```console
+120
+1 2 6 24 120
+```
+
+<a id="zip_and_agg"></a>
+## Zippered Iteration and Aggregation
+
+<a id="zippered_iteration"></a>
+### Zippered Iteration
+[Zippered Iteration](https://chapel-lang.org/docs/users-guide/base/zip.html#zippered-iteration) is simultaneously
+iterating over multiple iterables (usually arrays and ranges) with compatible domains.
+
+The syntax:
+```chapel
+forall (v1, v2, v3, ..., vn) in zip(A1, A2, A3, ..., An) {
+  // for loop iteration `j`, `vi` will refer to the `j`th element of `Ai`
+}
+```
+Since all `Ai` have compatible domains, we know all `vi` will be local to one another on any given loop iteration.
```

Comment:
Wait, do we know that they do?  We haven't seen the declarations of these arrays, have we?  This could be remedied by adding:

```chapel
var A1, A2, A3, ..., An: [something-something] int;
```